source/material_model/rheology/visco_plastic.cc

```diff
@@ -244,8 +248,6 @@ namespace aspect
                     // Explanation: When we compute the viscosity for the Stokes assembly
                     // the ve stresses stored in in.composition are the rotated, averaged,
                     // and advected stresses from the advection solve ($\tau^{t+\Delta t}$).
-                    // They should be the unaveraged, but advected stresses? How could we get
-                    // to those?
                     const double viscoelastic_strain_rate_invariant = elastic_rheology.calculate_viscoelastic_strain_rate(in.strain_rate[i],
                                                                       stress_old,
```

Comment:
@anne-glerum @bobmyhill - I wonder if here we should not be tau^t, but rather tau^(t+te)?

At present, we are using the updated strain rate, but the old viscoelastic stress. Maybe this could explain why the non-linear solvers are having so much trouble for problems with rapidly changing deformation fields?

Reply:
To follow, this would propagate into:
```
current_edot_ii = std::max(viscoelastic_strain_rate_invariant,
                                               min_strain_rate);
```

and then 
`double current_stress = 2. * viscosity_pre_yield * current_edot_ii;`

Reply:
When we need the viscosity for the Stokes assembly, in.composition holds tau^(t+dt), because it's after the Advection solve.